libcontainer/rootfs_linux.go

```diff
@@ -594,6 +595,37 @@ func createDevices(config *configs.Config) error {
 }
 
 func bindMountDeviceNode(dest string, node *configs.Device) error {
+	var st unix.Stat_t
+	if err := unix.Lstat(dest, &st); err != nil {
+		if !os.IsNotExist(err) {
+			return errors.Wrapf(err, "could not stat existing path for device node: %q", dest)
+		}
+	} else if os.FileMode(st.Mode)&os.ModeSymlink == os.ModeSymlink {
+		// If the path is a symlink, we remove it.
+		if err := unix.Unlink(dest); err != nil {
+			return errors.Wrapf(err, "device node %q already exists and is a symlink", dest)
+		}
+	} else {
+		switch os.FileMode(st.Mode) & os.ModeType {
+		case os.ModeDevice:
+			if node.Type != configs.BlockDevice {
+				return errors.Errorf("device node %q already exists but incorrect type %v (wanted block type)", dest, node.Type)
+			}
+		case os.ModeDevice | os.ModeCharDevice:
+			if node.Type != configs.CharDevice {
+				return errors.Errorf("device node %q already exists but incorrect type %v (wanted block type)", dest, node.Type)
+			}
+		default:
+			// NOP, happens on rootless
```

Comment:
This doesn't make sense (and is why I changed it in my suggestion) -- what you're saying here is "if it's any other kind of inode, ignore it unless `node.Major` and `node.Minor` are non-zero because if the inode isn't a device then it won't have a major/minor number". I don't understand why you would want to do that.

The actual bug (and it was unrelated to rootless containers) was that I forgot that `os.Mode*` has different values from the Linux ones. The actual switch statement should've looked like:

```
switch st.Mode & unix.S_IFMT {
case unix.S_IFBLK:
	gotType := configs.BlockDevice
case unix.S_IFCHR:
	gotType := configs.CharDevice
default:
	// I'm not sure if you want to also handle regular files as well?
	// I'm still a little confused what the semantics are supposed to be.
	return fmt.Errorf(...)
```

And you should change back the rest of the hunk too.

Reply:
Also (because of the misunderstanding I mentioned above), this code will ignore the type of the device because it will always go through the `default` case in the `switch`.